File: apps/web/lib/api/ip-patents.ts
import { z } from "zod";

import { apiRequest } from "./client";

const hash = z.string().regex(/^[a-f0-9]{64}$/);
const text = (max: number) => z.string().min(1).max(max).regex(/\S/)
  .refine((value) => !value.includes("\0"), "Text cannot contain a null character.");
const documentSource = z.object({
  kind: z.literal("document_version"), document_id: z.string().uuid(),
  document_version_id: z.string().uuid(), content_sha256: hash,
}).strict();
const registrySource = z.object({
  kind: z.literal("registry_snapshot"), snapshot_id: z.string().uuid(),
  raw_sha256: hash, normalized_sha256: hash,
}).strict();
export const patentSourceSchema = z.discriminatedUnion("kind", [documentSource, registrySource]);
export const patentFamilyFactsSchema = z.object({
  title: text(255), client_id: z.string().uuid(),
  disclosure_date: z.string().date(), disclosure_narrative: text(30_000),
  confidentiality: z.literal("restricted"), source: patentSourceSchema.nullable().optional(),
}).strict();
export const patentFamilySchema = z.object({
  id: z.string().uuid(), docket_id: z.string().uuid(), asset_id: z.string().uuid(),
  record_kind: z.literal("patent_family"), version: z.number().int().positive(),
  lifecycle_version: z.number().int().nonnegative(), lifecycle_status: z.string().max(32),
  is_active: z.boolean(), created_at: z.string().datetime({ offset: true }),
  updated_at: z.string().datetime({ offset: true }), facts: patentFamilyFactsSchema,
}).strict();
const familyListSchema = z.object({
  families: z.array(patentFamilySchema).max(100), next_cursor: z.string().max(512).nullable(),
}).strict();

export type PatentFamily = z.infer<typeof patentFamilySchema>;
export type PatentFamilyFacts = z.infer<typeof patentFamilyFactsSchema>;
export type PatentFamilyStatusScope = "active" | "terminal" | "all";

export async function fetchPatentFamilies(q: string, cursor?: string, signal?: AbortSignal, statusScope: PatentFamilyStatusScope = "active") {
  const query = new URLSearchParams({ limit: "50" });
  query.set("status_scope", statusScope);
  if (q) query.set("q", q);
  if (cursor) query.set("cursor", cursor);
  return familyListSchema.parse(await apiRequest<unknown>(`/api/ip/patents/families?${query}`, { signal }));
}

const lifecycleHistorySchema = z.object({
  events: z.array(z.object({
    id: z.string().uuid(), sequence: z.number().int().positive(),
    from_status: z.string().nullable(), to_status: z.string().nullable(),
    effective_at: z.string().datetime({ offset: true }), entered_at: z.string().datetime({ offset: true }),
    reason: z.string().nullable(), evidence_refs: z.array(z.string()).max(100),
  }).strict()).max(100),
  next_cursor: z.number().int().positive().nullable(),
}).strict();

export async function fetchPatentFamilyLifecycleHistory(id: string, cursor?: number, signal?: AbortSignal) {
  const query = new URLSearchParams({ limit: "25" });
  if (cursor !== undefined) query.set("cursor", String(cursor));
  return lifecycleHistorySchema.parse(await apiRequest<unknown>(
    `/api/ip/patents/families/${encodeURIComponent(id)}/lifecycle-history?${query}`, { signal },
  ));
}

export async function fetchPatentFamily(id: string, version?: number, signal?: AbortSignal) {
  const path = `/api/ip/patents/families/${encodeURIComponent(id)}`;
  return patentFamilySchema.parse(await apiRequest<unknown>(
    version ? `${path}/versions/${version}` : path, { signal },
  ));
}

export async function createPatentFamily(facts: PatentFamilyFacts, idempotencyKey: string) {
  return patentFamilySchema.parse(await apiRequest<unknown>("/api/ip/patents/families", {
    method: "POST", body: patentFamilyFactsSchema.parse(facts),
    headers: { "Idempotency-Key": idempotencyKey },
  }));
}

export async function correctPatentFamily(family: PatentFamily, facts: PatentFamilyFacts, reason: string) {
  return patentFamilySchema.parse(await apiRequest<unknown>(
    `/api/ip/patents/families/${encodeURIComponent(family.id)}/corrections`, {
      method: "POST", body: {
        expected_version: family.version, expected_lifecycle_version: family.lifecycle_version,
        reason, facts: patentFamilyFactsSchema.parse(facts),
      },
    },
  ));
}

export const patentApplicationKinds = [
  "provisional", "complete", "convention", "pct_international", "national_phase",
  "divisional", "patent_of_addition",
] as const;
export const patentIdentifierSchema = z.object({
  identifier_kind: z.enum(["application", "publication", "grant"]),
  raw_value: text(120), source: patentSourceSchema,
}).strict();
export const patentApplicationFactsSchema = z.object({
  title: text(255), application_kind: z.enum(patentApplicationKinds),
  jurisdiction: z.string().regex(/^[A-Z]{2}$/), office: text(80),
  filing_date: z.string().date().nullable(), publication_date: z.string().date().nullable(),
  identifiers: z.array(patentIdentifierSchema).max(20),
  source_pending_identifier_allocation: z.boolean(), source: patentSourceSchema,
}).strict().superRefine((facts, context) => {
  if (facts.filing_date && facts.publication_date && facts.publication_date < facts.filing_date) {
    context.addIssue({ code: "custom", path: ["publication_date"], message: "Publication cannot precede filing." });
  }
  if (facts.source_pending_identifier_allocation && facts.identifiers.some((row) => row.identifier_kind === "application")) {
    context.addIssue({ code: "custom", path: ["identifiers"], message: "An allocated application number cannot also be pending allocation." });
  }
  if (new Set(facts.identifiers.map((row) => JSON.stringify([row.identifier_kind, row.raw_value]))).size !== facts.identifiers.length) {
    context.addIssue({ code: "custom", path: ["identifiers"], message: "An identifier fact may appear only once." });
  }
});
export const patentApplicationSchema = z.object({
  id: z.string().uuid(), docket_id: z.string().uuid(), family_id: z.string().uuid(),
  record_kind: z.literal("patent_application"), version: z.number().int().positive(),
  lifecycle_version: z.number().int().nonnegative(), lifecycle_status: text(32), is_active: z.boolean(),
  created_at: z.string().datetime({ offset: true }), updated_at: z.string().datetime({ offset: true }),
  prosecution_phase: z.enum(["disclosure", "filing_preparation", "filed", "published", "examination_requested",
    "examination", "response_filed", "hearing", "granted", "refused", "withdrawn", "abandoned", "closed"]),
  facts: patentApplicationFactsSchema,
}).strict();
const applicationListSchema = z.object({
  applications: z.array(patentApplicationSchema).max(100), next_cursor: z.string().max(512).nullable(),
}).strict();
export type PatentApplication = z.infer<typeof patentApplicationSchema>;
export type PatentApplicationFacts = z.infer<typeof patentApplicationFactsSchema>;
export type PatentSource = z.infer<typeof patentSourceSchema>;

export async function fetchPatentApplications(familyId: string, cursor?: string, signal?: AbortSignal,
  statusScope: PatentFamilyStatusScope = "active", queryText = "") {
  const query = new URLSearchParams({ limit: "50", status_scope: statusScope });
  if (familyId) query.set("family_id", familyId);
  if (cursor) query.set("cursor", cursor);
  if (queryText) query.set("q", queryText);
  return applicationListSchema.parse(await apiRequest<unknown>(`/api/ip/patents/applications?${query}`, { signal }));
}

export async function fetchPatentApplication(id: string, version?: number, signal?: AbortSignal) {
  const path = `/api/ip/patents/applications/${encodeURIComponent(id)}`;
  return patentApplicationSchema.parse(await apiRequest<unknown>(version ? `${path}/versions/${version}` : path, { signal }));
}

export async function createPatentApplication(family: PatentFamily, facts: PatentApplicationFacts, idempotencyKey: string) {
  return patentApplicationSchema.parse(await apiRequest<unknown>("/api/ip/patents/applications", {
    method: "POST", headers: { "Idempotency-Key": idempotencyKey }, body: {
      family_id: family.id, expected_family_version: family.version,
      expected_family_lifecycle_version: family.lifecycle_version, facts: patentApplicationFactsSchema.parse(facts),
    },
  }));
}

export async function correctPatentApplication(application: PatentApplication, facts: PatentApplicationFacts, reason: string) {
  return patentApplicationSchema.parse(await apiRequest<unknown>(
    `/api/ip/patents/applications/${encodeURIComponent(application.id)}/corrections`, {
      method: "POST", body: { expected_version: application.version,
        expected_lifecycle_version: application.lifecycle_version, reason, facts: patentApplicationFactsSchema.parse(facts) },
    },
  ));
}

export async function fetchPatentApplicationLifecycleHistory(id: string, cursor?: number, signal?: AbortSignal) {
  const query = new URLSearchParams({ limit: "25" });
  if (cursor !== undefined) query.set("cursor", String(cursor));
  return lifecycleHistorySchema.parse(await apiRequest<unknown>(
    `/api/ip/patents/applications/${encodeURIComponent(id)}/lifecycle-history?${query}`, { signal },
  ));
}

export const patentPartyRoles = ["inventor", "applicant", "proprietor", "agent", "licensee"] as const;
export const patentPartyFactSchema = z.object({
  role: z.enum(patentPartyRoles), name: text(255), client_id: z.string().uuid().nullable(),
  address: z.object({
    address_lines: z.array(text(255)).min(1).max(5), city: text(120),
    region: text(120).nullable(), postal_code: text(40).nullable(),
    country_code: z.string().regex(/^[A-Z]{2}$/),
  }).strict(),
  effective_from: z.string().date(), effective_until: z.string().date().nullable(),
  source: patentSourceSchema,
}).strict().refine((fact) => !fact.effective_until || fact.effective_until >= fact.effective_from, {
  path: ["effective_until"], message: "The effective end cannot precede the start.",
});
export const patentPartySchema = z.object({
  id: z.string().uuid(), docket_id: z.string().uuid(), sequence: z.number().int().positive(),
  anchor_version: z.number().int().positive(), lifecycle_version: z.number().int().nonnegative(),
  supersedes_party_id: z.string().uuid().nullable(), is_current: z.boolean(),
  fact: patentPartyFactSchema, reason: text(1000).refine((value) => value.length >= 8),
  created_at: z.string().datetime({ offset: true }),
}).strict();
export const patentPartyPageSchema = z.object({
  docket_id: z.string().uuid(), collection_sequence: z.number().int().nonnegative(),
  parties: z.array(patentPartySchema).max(100), next_cursor: z.number().int().positive().nullable(),
}).strict();
export type PatentPartyFact = z.infer<typeof patentPartyFactSchema>;
export type PatentParty = z.infer<typeof patentPartySchema>;
export type PatentPartyPage = z.infer<typeof patentPartyPageSchema>;

export async function fetchPatentParties(docketId: string, history = false,
  continuation?: { cursor: number; sequence: number }, signal?: AbortSignal) {
  const id = z.string().uuid().parse(docketId);
  const query = new URLSearchParams({ limit: "25", history: String(history) });
  if (continuation) {
    query.set("cursor", String(z.number().int().positive().parse(continuation.cursor)));
    query.set("snapshot_sequence", String(z.number().int().nonnegative().parse(continuation.sequence)));
  }
  return patentPartyPageSchema.parse(await apiRequest<unknown>(
    `/api/ip/patents/dockets/${id}/parties?${query}`, { signal },
  ));
}

export async function createPatentParty(record: PatentFamily | PatentApplication, sequence: number,
  fact: PatentPartyFact, reason: string, supersedesId: string | null, idempotencyKey: string) {
  const request = z.object({
    expected_version: z.number().int().positive(), expected_lifecycle_version: z.number().int().nonnegative(),
    expected_party_sequence: z.number().int().nonnegative(),
    fact: patentPartyFactSchema, reason: text(1000).refine((value) => value.length >= 8),
    supersedes_party_id: z.string().uuid().nullable(),
  }).strict().parse({
    expected_version: record.version, expected_lifecycle_version: record.lifecycle_version,
    expected_party_sequence: sequence, fact, reason, supersedes_party_id: supersedesId,
  });
  return patentPartySchema.parse(await apiRequest<unknown>(
    `/api/ip/patents/dockets/${z.string().uuid().parse(record.docket_id)}/parties`, {
      method: "POST", headers: { "Idempotency-Key": idempotencyKey }, body: request,
    },
  ));
}

export const patentPriorityKinds = ["priority", "divisional_parent", "addition_parent", "national_phase_parent"] as const;
export const patentPrioritySchema = z.object({
  id: z.string().uuid(), canonical_relationship_id: z.string().uuid(),
  application_id: z.string().uuid(), parent_application_id: z.string().uuid(),
  current_application_title: text(255), current_parent_title: text(255),
  sequence: z.number().int().positive(), application_version: z.number().int().positive(),
  parent_version: z.number().int().positive(), lifecycle_version: z.number().int().nonnegative(),
  parent_lifecycle_version: z.number().int().nonnegative(), supersedes_priority_id: z.string().uuid().nullable(),
  withdrawn: z.boolean(), is_current: z.boolean(),
  review_flags: z.array(z.enum(["office_names_differ", "jurisdictions_differ", "filing_dates_incomplete",
    "priority_date_precedes_parent_filing"])).max(4),
  relation_kind: z.enum(patentPriorityKinds), priority_date: z.string().date(), source: patentSourceSchema,
  reason: text(1000).refine((value) => value.length >= 8), created_at: z.string().datetime({ offset: true }),
}).strict().refine((row) => row.application_id !== row.parent_application_id, "An application cannot link to itself.");
export const patentPriorityPageSchema = z.object({
  application_id: z.string().uuid(), collection_sequence: z.number().int().nonnegative(),
  priorities: z.array(patentPrioritySchema).max(100), next_cursor: z.number().int().positive().nullable(),
}).strict();
export const patentFamilyGraphSchema = z.object({
  family_id: z.string().uuid(), applications: z.array(patentApplicationSchema).max(100),
  priorities: z.array(patentPrioritySchema).max(500),
  applications_next_cursor: text(512).nullable(), priorities_next_cursor: text(512).nullable(),
  has_more_applications: z.boolean(), has_more_relationships: z.boolean(),
}).strict().refine((page) => page.has_more_applications === (page.applications_next_cursor !== null)
  && page.has_more_relationships === (page.priorities_next_cursor !== null), "Graph cursors must match continuation flags.")
  .refine((page) => page.applications.every((application) => application.family_id === page.family_id), "Graph applications must belong to this family.")
  .refine((page) => new Set(page.applications.map((application) => application.id)).size === page.applications.length
    && new Set(page.priorities.map((priority) => priority.id)).size === page.priorities.length, "Graph records cannot repeat identities.");
export type PatentPriority = z.infer<typeof patentPrioritySchema>;
export type PatentPriorityPage = z.infer<typeof patentPriorityPageSchema>;
export const patentPriorityCommandSchema = z.object({
  parent_application_id: z.string().uuid(), relation_kind: z.enum(patentPriorityKinds),
  priority_date: z.string().date(), source: patentSourceSchema,
  expected_application_version: z.number().int().positive(), expected_parent_version: z.number().int().positive(),
  expected_lifecycle_version: z.number().int().nonnegative(), expected_parent_lifecycle_version: z.number().int().nonnegative(),
  expected_priority_sequence: z.number().int().nonnegative(), supersedes_priority_id: z.string().uuid().nullable(),
  withdrawn: z.boolean(), reason: text(1000).refine((value) => value.length >= 8),
}).strict().refine((command) => !command.withdrawn || command.supersedes_priority_id !== null,
  "A withdrawal must identify the recorded relationship.");
export type PatentPriorityCommand = z.infer<typeof patentPriorityCommandSchema>;

export async function fetchPatentPriorities(applicationId: string, history = false,
  continuation?: { cursor: number; sequence: number }, signal?: AbortSignal) {
  const query = new URLSearchParams({ limit: "25", history: String(history) });
  if (continuation) {
    query.set("cursor", String(z.number().int().positive().parse(continuation.cursor)));
    query.set("snapshot_sequence", String(z.number().int().nonnegative().parse(continuation.sequence)));
  }
  return patentPriorityPageSchema.parse(await apiRequest<unknown>(
    `/api/ip/patents/applications/${z.string().uuid().parse(applicationId)}/priorities?${query}`, { signal },
  ));
}

export async function createPatentPriority(applicationId: string, command: PatentPriorityCommand, idempotencyKey: string) {
  return patentPrioritySchema.parse(await apiRequest<unknown>(
    `/api/ip/patents/applications/${z.string().uuid().parse(applicationId)}/priorities`, {
      method: "POST", headers: { "Idempotency-Key": idempotencyKey }, body: patentPriorityCommandSchema.parse(command),
    },
  ));
}

export async function fetchPatentFamilyGraph(familyId: string, applicationsCursor?: string, prioritiesCursor?: string,
  signal?: AbortSignal) {
  const query = new URLSearchParams({ application_limit: "25", priority_limit: "25" });
  if (applicationsCursor) query.set("applications_cursor", applicationsCursor);
  if (prioritiesCursor) query.set("priorities_cursor", prioritiesCursor);
  return patentFamilyGraphSchema.parse(await apiRequest<unknown>(
    `/api/ip/patents/families/${z.string().uuid().parse(familyId)}/graph?${query}`, { signal },
  ));
}

File: apps/api/src/caseops_api/schemas/ip_patents.py
"""PAT-2026-09-06.1 typed facts; these contracts do not activate patent intake.

Source pins reference existing document/registry owners. A schema-valid pin is
not authorization or verified legal evidence: writers must resolve it under the
current tenant, docket access and immutable content hash before persistence.
"""

from __future__ import annotations

from datetime import date
from typing import Annotated, Literal, Self
from uuid import UUID

from pydantic import (
    AfterValidator,
    AwareDatetime,
    BaseModel,
    ConfigDict,
    Field,
    StrictBool,
    StringConstraints,
    model_validator,
)

PatentApplicationKind = Literal[
    "provisional",
    "complete",
    "convention",
    "pct_international",
    "national_phase",
    "divisional",
    "patent_of_addition",
]
PatentProsecutionPhase = Literal[
    "disclosure",
    "filing_preparation",
    "filed",
    "published",
    "examination_requested",
    "examination",
    "response_filed",
    "hearing",
    "granted",
    "refused",
    "withdrawn",
    "abandoned",
    "closed",
]
PatentIdentifierKind = Literal["application", "publication", "grant"]
PatentPartyRole = Literal["inventor", "applicant", "proprietor", "agent", "licensee"]
PatentRelationKind = Literal[
    "priority",
    "divisional_parent",
    "addition_parent",
    "national_phase_parent",
]
PatentPriorityReviewFlag = Literal[
    "office_names_differ",
    "jurisdictions_differ",
    "filing_dates_incomplete",
    "priority_date_precedes_parent_filing",
]
PatentProceedingKind = Literal[
    "pre_grant_opposition",
    "post_grant_opposition",
    "revocation",
    "compulsory_licence",
]


def _no_null_character(value: str) -> str:
    if "\x00" in value:
        raise ValueError("Text cannot contain a null character.")
    return value


NonBlankText = Annotated[str, StringConstraints(pattern=r"\S"), AfterValidator(_no_null_character)]
CountryCode = Annotated[str, StringConstraints(pattern=r"^[A-Z]{2}$")]
Sha256 = Annotated[str, StringConstraints(pattern=r"^[a-f0-9]{64}$")]
PositiveVersion = Annotated[int, Field(strict=True, ge=1)]


class PatentContract(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class PatentDocumentSource(PatentContract):
    kind: Literal["document_version"] = "document_version"
    document_id: UUID
    document_version_id: UUID
    content_sha256: Sha256


class PatentRegistrySource(PatentContract):
    kind: Literal["registry_snapshot"] = "registry_snapshot"
    snapshot_id: UUID
    raw_sha256: Sha256
    normalized_sha256: Sha256


PatentSourcePin = Annotated[
    PatentDocumentSource | PatentRegistrySource, Field(discriminator="kind")
]


class PatentIdentifierFact(PatentContract):
    identifier_kind: PatentIdentifierKind
    raw_value: NonBlankText = Field(min_length=1, max_length=120)
    source: PatentSourcePin


class PatentAddressSnapshot(PatentContract):
    address_lines: tuple[Annotated[NonBlankText, Field(max_length=255)], ...] = Field(
        min_length=1,
        max_length=5,
    )
    city: NonBlankText = Field(max_length=120)
    region: NonBlankText | None = Field(default=None, max_length=120)
    postal_code: NonBlankText | None = Field(default=None, max_length=40)
    country_code: CountryCode


class PatentPartyFact(PatentContract):
    role: PatentPartyRole
    name: NonBlankText = Field(max_length=255)
    client_id: UUID | None = None
    address: PatentAddressSnapshot
    effective_from: date
    effective_until: date | None = None
    source: PatentSourcePin

    @model_validator(mode="after")
    def ordered_effective_range(self) -> Self:
        if self.effective_until is not None and self.effective_until < self.effective_from:
            raise ValueError("A party's effective end cannot precede its start.")
        return self


class PatentFamilyFacts(PatentContract):
    title: NonBlankText = Field(max_length=255)
    client_id: UUID
    disclosure_date: date
    disclosure_narrative: NonBlankText = Field(max_length=30_000)
    confidentiality: Literal["restricted"] = "restricted"
    source: PatentSourcePin | None = None


class PatentFamilyCreateRequest(PatentFamilyFacts):
    pass


class PatentFamilyCorrectionRequest(PatentContract):
    expected_version: PositiveVersion
    expected_lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    reason: NonBlankText = Field(min_length=8, max_length=1000)
    facts: PatentFamilyFacts


class PatentFamilyLifecycleEvent(PatentContract):
    id: UUID
    sequence: PositiveVersion
    from_status: str | None
    to_status: str | None
    effective_at: AwareDatetime
    entered_at: AwareDatetime
    reason: str | None
    evidence_refs: list[str] = Field(max_length=100)


class PatentFamilyLifecycleHistory(PatentContract):
    events: list[PatentFamilyLifecycleEvent] = Field(max_length=100)
    next_cursor: PositiveVersion | None


class PatentApplicationFacts(PatentContract):
    title: NonBlankText = Field(max_length=255)
    application_kind: PatentApplicationKind
    jurisdiction: CountryCode
    office: NonBlankText = Field(max_length=80)
    filing_date: date | None = None
    publication_date: date | None = None
    identifiers: tuple[PatentIdentifierFact, ...] = Field(default=(), max_length=20)
    source_pending_identifier_allocation: StrictBool = False
    source: PatentSourcePin

    @model_validator(mode="after")
    def consistent_recorded_facts(self) -> Self:
        if (
            self.publication_date is not None
            and self.filing_date is not None
            and self.publication_date < self.filing_date
        ):
            raise ValueError("Publication cannot precede the recorded filing date.")
        if self.source_pending_identifier_allocation and any(
            identifier.identifier_kind == "application" for identifier in self.identifiers
        ):
            raise ValueError("An allocated application number cannot also be pending allocation.")
        raw_keys = {(row.identifier_kind, row.raw_value) for row in self.identifiers}
        if len(raw_keys) != len(self.identifiers):
            raise ValueError("An identifier fact may appear only once.")
        return self


class PatentApplicationCreateRequest(PatentContract):
    family_id: UUID
    expected_family_version: PositiveVersion
    expected_family_lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    facts: PatentApplicationFacts


class PatentApplicationCorrectionRequest(PatentContract):
    expected_version: PositiveVersion
    expected_lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    reason: NonBlankText = Field(min_length=8, max_length=1000)
    facts: PatentApplicationFacts


class PatentPriorityCreateRequest(PatentContract):
    parent_application_id: UUID
    expected_application_version: PositiveVersion
    expected_parent_version: PositiveVersion
    expected_lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    expected_parent_lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    expected_priority_sequence: Annotated[int, Field(strict=True, ge=0)]
    supersedes_priority_id: UUID | None = None
    withdrawn: StrictBool = False
    relation_kind: PatentRelationKind
    priority_date: date
    source: PatentSourcePin
    reason: NonBlankText = Field(min_length=8, max_length=1000)

    @model_validator(mode="after")
    def withdrawal_requires_predecessor(self) -> Self:
        if self.withdrawn and self.supersedes_priority_id is None:
            raise ValueError("Withdrawing a recorded link requires its current predecessor.")
        return self


class PatentPartyCreateRequest(PatentContract):
    expected_version: PositiveVersion
    expected_lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    expected_party_sequence: Annotated[int, Field(strict=True, ge=0)]
    supersedes_party_id: UUID | None = None
    reason: NonBlankText = Field(min_length=8, max_length=1000)
    fact: PatentPartyFact


class PatentAnchorRecord(PatentContract):
    id: UUID
    docket_id: UUID
    version: PositiveVersion
    lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    lifecycle_status: NonBlankText = Field(max_length=32)
    is_active: StrictBool
    created_at: AwareDatetime
    updated_at: AwareDatetime


class PatentFamilyRecord(PatentAnchorRecord):
    record_kind: Literal["patent_family"] = "patent_family"
    asset_id: UUID
    facts: PatentFamilyFacts


class PatentApplicationRecord(PatentAnchorRecord):
    record_kind: Literal["patent_application"] = "patent_application"
    family_id: UUID
    prosecution_phase: PatentProsecutionPhase
    facts: PatentApplicationFacts


class PatentPartyRecord(PatentContract):
    id: UUID
    docket_id: UUID
    sequence: PositiveVersion
    anchor_version: PositiveVersion
    lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    supersedes_party_id: UUID | None = None
    is_current: StrictBool
    fact: PatentPartyFact
    reason: NonBlankText = Field(min_length=8, max_length=1000)
    created_at: AwareDatetime


class PatentPartyListResponse(PatentContract):
    docket_id: UUID
    collection_sequence: Annotated[int, Field(strict=True, ge=0)]
    parties: tuple[PatentPartyRecord, ...] = Field(max_length=100)
    next_cursor: PositiveVersion | None = None


class PatentPriorityRecord(PatentContract):
    id: UUID
    canonical_relationship_id: UUID
    application_id: UUID
    parent_application_id: UUID
    current_application_title: NonBlankText = Field(max_length=255)
    current_parent_title: NonBlankText = Field(max_length=255)
    sequence: PositiveVersion
    application_version: PositiveVersion
    parent_version: PositiveVersion
    lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    parent_lifecycle_version: Annotated[int, Field(strict=True, ge=0)]
    supersedes_priority_id: UUID | None = None
    withdrawn: StrictBool
    is_current: StrictBool
    review_flags: tuple[PatentPriorityReviewFlag, ...] = Field(max_length=4)
    relation_kind: PatentRelationKind
    priority_date: date
    source: PatentSourcePin
    reason: NonBlankText = Field(min_length=8, max_length=1000)
    created_at: AwareDatetime

    @model_validator(mode="after")
    def no_self_priority(self) -> Self:
        if self.application_id == self.parent_application_id:
            raise ValueError("An application cannot claim priority from itself.")
        return self


class PatentPriorityListResponse(PatentContract):
    application_id: UUID
    collection_sequence: Annotated[int, Field(strict=True, ge=0)]
    priorities: tuple[PatentPriorityRecord, ...] = Field(max_length=100)
    next_cursor: PositiveVersion | None = None


class PatentFamilyListResponse(PatentContract):
    families: tuple[PatentFamilyRecord, ...] = Field(max_length=100)
    next_cursor: str | None = Field(default=None, max_length=512)


class PatentApplicationListResponse(PatentContract):
    applications: tuple[PatentApplicationRecord, ...] = Field(max_length=100)
    next_cursor: str | None = Field(default=None, max_length=512)


class PatentFamilyGraphResponse(PatentContract):
    family_id: UUID
    applications: tuple[PatentApplicationRecord, ...] = Field(max_length=100)
    priorities: tuple[PatentPriorityRecord, ...] = Field(max_length=500)
    applications_next_cursor: NonBlankText | None = Field(default=None, max_length=512)
    priorities_next_cursor: NonBlankText | None = Field(default=None, max_length=512)
    has_more_applications: StrictBool
    has_more_relationships: StrictBool

    @model_validator(mode="after")
    def explicit_independent_continuation(self) -> Self:
        if self.has_more_applications != (self.applications_next_cursor is not None):
            raise ValueError("An incomplete application page requires its own continuation.")
        if self.has_more_relationships != (self.priorities_next_cursor is not None):
            raise ValueError("An incomplete priority page requires its own continuation.")
        if any(row.family_id != self.family_id for row in self.applications):
            raise ValueError("A family page cannot contain another family's applications.")
        for rows in (self.applications, self.priorities):
            if len({row.id for row in rows}) != len(rows):
                raise ValueError("A graph page cannot repeat the same record identity.")
        return self
